### DISCUSSION TITLE
Megadesk Companion Flashing

### DISCUSSION THREAD

**Discussion by medoix** [gcormier/megadesk#138] [Q&A]
I found this discussion (https://github.com/gcormier/megadesk/discussions/106) that links to a DIY file with commands on flashing the Megadesk but there appears to be no details on flashing the Companion board. 

Is there some details on flashing options (CLI or Web) for the companion board after shorting GND and GP0 pins for the first time?

**Comment by gcormier** [comment]
I typically use the command line to flash. Ideally, getting the most recent `esphome` version is preferred via `pip`.

It should be as simple as

1. Download [megadesk-companion.yaml](https://github.com/gcormier/megadesk/blob/master/esphome/megadesk-companion.yaml)
2. `esphome compile megadesk-companion.yaml`
3. `esphome upload megadesk-companion.yaml`

The steps via GUI or homeassistant should be very similar.

**Comment by gcormier** [comment]
## Flashing using docker
Benefit is you don't need to mess with python at all, you'll get the latest esphome version.

1. Find out the serial port that represents megadesk-companion. I'll assume /dev/ttyUSB0 but it might be different for you.

1. Have [megadesk-companion.yaml](https://github.com/gcormier/megadesk/blob/master/esphome/megadesk-companion.yaml) and your `secrets.yaml` file in the current working directory.

`docker run -v ./:/myconfig --device /dev/ttyUSB0 esphome/esphome upload /myconfig/megadesk-companion.yaml --device /dev/ttyUSB0`

- Note the first occurence of `--device` tells docker to mount that device inside the container. The second occurence of `--device` is telling esphome to upload the compiled firmware via that device.

3. If you want to upload it via WiFi (after initial config), you would use 

`docker run -v ./:/myconfig esphome/esphome upload /myconfig/megadesk-companion.yaml --device 192.168.x.y`

**Comment by medoix** [comment]
> I typically use the command line to flash. Ideally, getting the most recent `esphome` version is preferred via `pip`.
> 
> It should be as simple as
> 
>     1. Download [megadesk-companion.yaml](https://github.com/gcormier/megadesk/blob/master/esphome/megadesk-companion.yaml)
> 
>     2. `esphome compile megadesk-companion.yaml`
> 
>     3. `esphome upload megadesk-companion.yaml`
> 
> 
> The steps via GUI or homeassistant should be very similar.

This was all i needed, for some reason after all of the googling and reading on the esphome website i went down many rabbit holes instead of simply doing a ```pip install esphome``` 👍 

companion is flashed and connected to the WiFi, thanks!

**Comment by raleighlittles** [comment]
I haven't seen this mentioned anywhere but you need to also edit the "wifi" section of `megadesk-companion.yaml` for your router's SSID and password

![image](https://github.com/gcormier/megadesk/assets/2137201/b31477e5-54fc-4a25-9a00-60d57a3fbe75)

**Comment by webmediart-github** [comment]
I just compiled a small documentation with a bit more detail for those who are not that familiar with the stuff happening. Hope it helps, cheers!

## Prerequisites
- Download/Install [CP210x drivers]](https://www.silabs.com/developers/usb-to-uart-bridge-vcp-drivers?tab=downloads) if required
- Download [megadesk-companion.yaml](https://github.com/gcormier/megadesk/blob/master/esphome/megadesk-companion.yaml)
- Download [megadesk.h](https://github.com/gcormier/megadesk/blob/master/esphome/megadesk.h)
- Create `secrets.yaml`

		  ```yaml
		  wifi_ssid: REPLACE_WITH_WIFI-SSID
		  wifi_password: REPLACE_WITH_WIFI-PASSWORD
		  ```
- Install Python (needed for #ESPHome)
- Install #ESPHome Add-on in Homeassistant
- Install #ESPHome on flashing machine (e.g. via Terminal `pip install esphome`)


## Steps
1. Start terminal and change directory into folder with `megadesk-companion.yaml`, `secrets.yaml` and `megadesk.h` in it

    - Commands on Windows-Terminal
      - Change directory: `cd` followed by the path you want to change into (`cd ..` will jump back/up one level)
      - List files and directories of current directory: `dir`
2. Connect the Companion-PCB via USB-C with the flashing machine, while pressing the button on the PCB

	- pressing the button while connecting is only needed for the first time flashing
	- successful if cmd `esptool.py chip_id` does not throw a fatal error
3. Run cmd `esphome compile megadesk-companion.yaml`

4. Run cmd `esphome upload megadesk-companion.yaml`
       - chose port 1 when promted (input "1" and confirm with return)
